### ISSUE TITLE
I don't get performance gain by using Intel-Engine on VPU Myriad X for OpenCV-dnn module

### ISSUE THREAD

**Issue by AlexeyAB** [opencv/opencv#15023]
##### System information (version)

- OpenCV => 4.1.0
- Operating System / Platform => Ubuntu 16.04 x86_64
- Compiler => g++ 5.4.0
- CPU Atom E3845 + VPU Myriad X

##### Detailed description

I don't get any significant performance gain by using Intel-Engine on VPU Myriad X for OpenCV-dnn module.

I get the same performance with CPU Atom E3845 and VPU Myriad X:
* **0.3 FPS** with `backend=Intel-Engine and target=VPU`
* **0.25 FPS** with `backend=OpenCV-impl target=CPU`

##### Steps to reproduce

I compiled OpenVINO 2019 R1. 

I compiled OpenCV 4.1.0 with Intel's Deep Learning Inference Engine (DL IE) as described here: https://github.com/opencv/opencv/wiki/Intel's-Deep-Learning-Inference-Engine-backend

I modified this example (minor changes, but you can use this original file with the same effect): https://github.com/opencv/opencv/blob/master/samples/dnn/object_detection.cpp
to this: `ocv_yolo.cpp` that is in attached [repr.zip](https://github.com/opencv/opencv/files/3383463/repr.zip)

And run this command: 
`./ocv_yolo --config=yolov3.cfg --model=yolov3.weights --input=test.mp4 --width=416 --height=416 --classes=coco.names --scale=0.00392 --rgb --backend=2 --targer=3 `

As known:
* backend 0(auto), 1(Halide), 2(Intel-Engine), 3(OpenCV-impl)
* target 0(CPU), 1(OpenCL), 2(OpenCV-FP16), 3(VPU)

But I get the same performance 
* **0.3 FPS** with `backend=2 and target=3`
* **0.25 FPS** with `backend=3 target=0`

While by using OpenVINO example I get about **4.3 FPS** on the same model. And it works successfully on Myriad X (VPU) for `frozen_yolov3.pb` TensorFlow-model that I got from `yolov3.cfg/weights` Darknet-model with **4.3 FPS** by using this example from OpenVINO: https://github.com/opencv/open_model_zoo/blob/master/demos/object_detection_demo_yolov3_async/main.cpp

Source code to reproduce the problem - that uses OpenCV-dnn module for Yolov3 detection, required any `test.mp4` file: [repr.zip](https://github.com/opencv/opencv/files/3383463/repr.zip)
Also required: 
* https://pjreddie.com/media/files/yolov3.weights
* https://raw.githubusercontent.com/pjreddie/darknet/master/cfg/yolov3.cfg

**Comment by alalek** [comment]
> targer=3

typo?

> get performance

What is about "Performance per watt"?

Atom - about 9W
Myriad - about 1W

**Comment by AlexeyAB** [comment]
@alalek Hi, Thanks!

> targer=3
> 
> typo?
> 

Yes :)
`./ocv_yolo --config=yolov3.cfg --model=yolov3.weights --input=test.mp4 --width=416 --height=416 --classes=coco.names --scale=0.00392 --rgb --backend=2 --target=3`

* With `--backend=2 --target=3` I get **1.7 FPS** with **OpenCV-dnn**

While with **OpenVINO** example https://github.com/opencv/open_model_zoo/blob/master/demos/object_detection_demo_yolov3_async/main.cpp 
* I get **2.3 FPS** in sync-mode 
* and **4.3 FPS** in async-mode


1. Does OpenCV-dnn with DL IE (VPU Myriad) uses FP16 precision automatically?

2. Is it possible to achieve approximately the same performance with OpenCV-dnn as with OpenVINO? (~2 - 4 FPS) Or does the OpenCV-dnn module with DL Inference Engine have large overheads?

----

Also initially when I run this sample, I always get several errors, and after 10-15 sec it goes well:
![image](https://user-images.githubusercontent.com/4096485/61080069-e6d8e080-a42c-11e9-8c56-27026f73c506.png)

**Comment by AlexeyAB** [comment]
Also I tried to use `--async=1` flag but it gives me this error:

`./ocv_yolo --config=yolov3.cfg --model=yolov3.weights --input=test.mp4 --width=416 --height=416 --classes=coco.names --scale=0.00392 --rgb --backend=2 --target=3 --async=1`

![image](https://user-images.githubusercontent.com/4096485/61080753-6dda8880-a42e-11e9-8b82-7dd0421bf16f.png)

**Comment by alalek** [comment]
> async=1

make no sense (slower than sync mode)

Try `--async=2` (see #14516, compile example with `-std=c++11` and apply latest changes from OpenCV "master" branch)

Mentioned message is from [here](https://github.com/opencv/opencv/blob/f663e8f903645a3dd66f6833f63717b86e861d77/samples/dnn/object_detection.cpp#L269-L270) and it is not correct (/cc @dkurt).

**Comment by dkurt** [comment]
The problem is that OpenVINO's Region layer is different from OpenCV's Region. So the network is divided on multiple ones and most of the time takes data transfer between VPU and fallback to Region layer.

It's known problem and as a workaround you can run Region separately (remove them from cfg). But definitely the best solution is to enable Region with IE backend and add new postprocessing stage for OpenVINO's Region. /cc @l-bat

**Comment by AlexeyAB** [comment]
@dkurt 

> But definitely the best solution is to enable Region with IE backend and add new postprocessing stage for OpenVINO's Region.

How can I do this, or is it an internal OpenCV+OpenVINO issue which will be solved later?
Why they did 2 different implementations (opencv and openvino) of one Region layer?
And can we use several NCS2 / Myriad X chips to process one neural network, should I just create two networks and execute alternately asynchronously? (As I do it in OpenVINO)

**Comment by dkurt** [comment]
@AlexeyAB, Comparing both implementations we can find that OpenCV's Region loop

https://github.com/opencv/opencv/blob/8b52eabd48195c42b12062735b1ffb0e911d389b/modules/dnn/src/layers/region_layer.cpp#L269-L291

in OpenVINO is outside the deep learning network but a part of sample (postprocessing):

https://github.com/opencv/open_model_zoo/blob/e4d1e85b3ed42372ce3514b98a15f51da94387d6/demos/object_detection_demo_yolov3_async/main.cpp#L122-L172

Looks a bit confusing because this is actually a part of layer in Darknet (perhaps they were driven by TensorFlow port instead).

We want to experiment with Region layer a bit and provide an update for object detection layer with one more postprocessing for OpenVINO's Region (after merge of https://github.com/opencv/opencv/pull/14780 it'll be hidden from samples at all).

**Comment by AlexeyAB** [comment]
@dkurt Thanks!

So OpenVINO just performs only logistic activation, and then we should get detection and do non-maximum-supression in postprocessing code.
Also in OpenVINO example the anchors are hardcoded (that is not correct), which should be different for different models: yolov3, yolov3-tiny, and many new models: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968

> So the network is divided on multiple ones and most of the time takes data transfer between VPU and fallback to Region layer.

Does this issue occur only for OpenCV-dnn, not for OpenVINO?

----

PS
Also do you think about Convolutional-LSTM layers implementation (recurrent sequence-to-sequence), which seems to give a huge increase in accuracy for detection on video (instead of individual images) ~1.5x higher accuracy: https://github.com/AlexeyAB/darknet/issues/3114#issue-441429452

**Comment by dkurt** [comment]
>  Also in OpenVINO example the anchors are hardcoded (that is not correct), which should be different for different models: yolov3, yolov3-tiny, and many new models: AlexeyAB/darknet#3114 (comment)

There is a comming fix for masks (2019R2) for these OpenVINO samples

> Does this issue occur only for OpenCV-dnn, not for OpenVINO?

Only for OpenCV-dnn. We'll try to fix it so native Darknet models could be ported to Myriad entirely.

---
>  Also do you think about Convolutional-LSTM layers implementation (recurrent sequence-to-sequence), which seems to give a huge increase in accuracy for detection on video (instead of individual images) ~1.5x higher accuracy: AlexeyAB/darknet#3114 (comment)

Let's open a separate thread for this issue :) LSTM is implemented so we need just check that it matches your version and add modify parser.

**Comment by AlexeyAB** [comment]
@dkurt Thanks, I added issue for conv-LSTM: https://github.com/opencv/opencv/issues/15033

**Comment by dkurt** [comment]
@AlexeyAB, Please try the changes proposed in https://github.com/opencv/opencv/pull/15063. It gives me about 390ms against 462ms on master branch for YOLOv3 (NCS2).

This PR also lets to run YOLO models asynchronously except YOLOv3 with multiple outputs. So for now you can test YOLOv3 with sync mode or async but only with output of a single layer.

I'll open PR with `forwardAsync` for multiple layers separately.
